06.bowling_object/frame.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+require_relative 'shot'
+
+class Frame
+  LAST_FRAME = 9
+  MAX_SHOTS_IN_LAST_FRAME = 3
+  MAX_SHOTS_IN_REGULAR_FRAME = 2
+
+  attr_reader :scores
+
+  def initialize(frame_idx)
+    @scores = []
+    @frame_idx = frame_idx
+  end
+
+  def add_shot(marks)
+    marks.each do |mark|
+      @scores << Shot.new(mark)
+      break if complete?
+    end
+    @scores.size
+  end
+
+  def score(frames)
+    @frame_idx != LAST_FRAME ? @scores.sum(&:score) + bonus_score(frames) : @scores.sum(&:score)
+  end
+
+  private
+
+  def complete?
+    if @frame_idx != LAST_FRAME
+      strike? || @scores.size == MAX_SHOTS_IN_REGULAR_FRAME
+    else
+      @scores.size == (strike? || spare? ? MAX_SHOTS_IN_LAST_FRAME : MAX_SHOTS_IN_REGULAR_FRAME)
+    end
+  end
+
+  def bonus_score(frames)
+    next_frame = frames[@frame_idx + 1]
+    second_next_frame = frames[@frame_idx + 2]
+    if strike?
+      strike_bonus(next_frame, second_next_frame)
+    elsif spare?
+      spare_bonus(next_frame)
+    else
+      0
+    end
+  end
+
+  def strike?
+    @scores[0].strike?
+  end
+
+  def spare?
+    !strike? && @scores[0..1].sum(&:score) == 10
+  end
+
+  def strike_bonus(next_frame, second_next_frame)
+    bonus = next_frame.scores[0].score
+    next_frame.scores[1] ? bonus + next_frame.scores[1].score : bonus + second_next_frame.scores[0].score
```

Comment:
```suggestion
    next_frame.scores[0].score + (next_frame.scores[1] || second_next_frame.scores[0]).score
```

ややトリッキーですがこんな書き方もできますね

Reply:
こんな書き方もできてしまうのですね！